File: modules/agent/src/models/Conversation.py
from sqlalchemy import Column, Text, DateTime, BigInteger, ForeignKey, CheckConstraint, func
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import relationship
from models import Base


class Conversation(Base):
    """Chat conversation metadata."""

    __tablename__ = "Conversation"

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    tenant_id = Column(BigInteger, ForeignKey("Tenant.id"), nullable=False)
    user_id = Column(BigInteger, ForeignKey("User.id"), nullable=False)
    thread_id = Column(UUID(as_uuid=True), nullable=False, unique=True)
    title = Column(Text, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)
    archived_at = Column(DateTime(timezone=True), nullable=True)

    messages = relationship("ConversationMessage", back_populates="conversation", cascade="all, delete-orphan")


class ConversationMessage(Base):
    """Individual message within a conversation."""

    __tablename__ = "Conversation_Message"

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    conversation_id = Column(BigInteger, ForeignKey("Conversation.id", ondelete="CASCADE"), nullable=False)
    role = Column(Text, nullable=False)
    content = Column(Text, nullable=False)
    token_usage = Column(JSONB, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    conversation = relationship("Conversation", back_populates="messages")

    __table_args__ = (
        CheckConstraint(role.in_(["user", "assistant"]), name="ck_conversation_message_role"),
    )

File: modules/agent/internal/models/Conversation.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/datatypes"
)

type Conversation struct {
	ID          int64      `gorm:"primaryKey;autoIncrement" json:"id"`
	TenantID    int64      `gorm:"index;not null" json:"tenant_id"`
	UserID      int64      `gorm:"not null" json:"user_id"`
	ThreadID    uuid.UUID  `gorm:"type:uuid;uniqueIndex;not null" json:"thread_id"`
	Title       *string    `json:"title"`
	CreatedAt   time.Time  `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt   time.Time  `gorm:"autoUpdateTime" json:"updated_at"`
	ArchivedAt  *time.Time `json:"archived_at"`
	CreatedByID int64      `gorm:"not null" json:"created_by_id"`
	UpdatedByID int64      `gorm:"not null" json:"updated_by_id"`
}

func (Conversation) TableName() string {
	return "Conversation"
}

type ConversationMessage struct {
	ID             int64          `gorm:"primaryKey;autoIncrement" json:"id"`
	ConversationID int64          `gorm:"index;not null" json:"conversation_id"`
	Role           string         `gorm:"not null" json:"role"` // "user" or "assistant"
	Content        string         `gorm:"not null" json:"content"`
	TokenUsage     datatypes.JSON `json:"token_usage,omitempty"`
	CreatedAt      time.Time      `gorm:"autoCreateTime" json:"created_at"`
}

func (ConversationMessage) TableName() string {
	return "Conversation_Message"
}
